### PR TITLE
1-Improvements

### PR DESCRIPTION
**Pull request by jumferreira**
#1

### CODE REVIEW

File: src/App.vue [hunk 1]
```diff
@@ -122,130 +148,161 @@
...
-                                @click="closeSectionModal()"
+                                class="
+                                    bg-primary hover:bg-primary-light
+                                    text-white font-bold py-2 px-4 rounded
+                                    border-b-4 border-primary-dark hover:border-primary
+                                "
+                                @click.prevent="submitSectionModal(sectionModal.sectionKey)"
                             >
-                                <span><i class="fas fa-times"></i></span>
+                                Add Section
```
Comment (gabrielboliveira) [inline src/App.vue]: Prefer using `v-text="'Add Section'"`

File: src/App.vue [hunk 2]
```diff
@@ -122,130 +148,161 @@
...
                     </div>
+
+                    <form class="">
+                        <div class="border-b border-b-2 border-primary py-2 mt-3">
+                            <textarea
+                                class="appearance-none bg-transparent border-none w-full text-grey-darker py-1 px-2"
+                                rows="9"
+                                v-model="transferModal.data"
+                            >
+                            </textarea>
```
Comment (gabrielboliveira) [inline src/App.vue]: move the closing tag to the same line as `>`

File: src/App.vue [hunk 3]
```diff
@@ -370,19 +433,31 @@ export default {
...
 
-        submitItemModal() {
-            this.sections[this.itemModal.sectionKey].items.push({
-                title: this.itemModal.title,
-                icon: this.itemModal.icon,
-                image: this.itemModal.image,
-                url: this.itemModal.url,
-            });
+        submitItemModal(sectionKey, itemKey) {
+            if (itemKey !== undefined) {
```
Comment (gabrielboliveira) [inline src/App.vue]: Use direct return:

```js
if (...) {
    return;
} 
// else block without the else
```

File: src/App.vue [hunk 4]
```diff
@@ -402,21 +478,60 @@ export default {
...
+            this.openImportExport();
+        },
+
+        openImportExport () {
+            this.$modal.show('transferInfos');
+
+            this.tranferTitle = this.isImport ? 'Import' : 'Export';
+            let local = window.localStorage;
+
+            let storage = JSON.stringify(local.App).replace(/\\/g, "");
```
Comment (gabrielboliveira) [inline src/App.vue]:
```js
let local = window.localStorage,
    storage = JSON.stringify(local.App).replace(/\\/g, "");
```

File: src/App.vue [hunk 5]
```diff
@@ -429,6 +544,50 @@ export default {
...
+                       image: item.image || '',
+                       url: item.url,
+                   });
+                });
+
+                sectionsToImport.push({
+                    title: section.title,
+                    items: itemsToImport,
+                });
+            });
```
Comment (gabrielboliveira) [inline src/App.vue]: Instead of using forEach and an external array to push items in it, use lodash's map function

File: src/App.vue [hunk 6]
```diff
@@ -370,19 +433,31 @@ export default {
...
+                this.sections[sectionKey].items[itemKey].image = this.itemModal.image;
+                this.sections[sectionKey].items[itemKey].url = this.itemModal.url;
+            } else {
+                this.sections[this.itemModal.sectionKey].items.push({
+                    title: this.itemModal.title,
+                    icon: this.itemModal.icon,
+                    image: this.itemModal.image,
+                    url: this.itemModal.url,
+                });
+            }
```
Comment (wesleyhf) [inline src/App.vue]: try this code:

```javascript
let section = this.sections[sectionKey];
let item = {
    title: this.itemModal.title,
    icon: this.itemModal.icon,
    image: this.itemModal.image,
    url: this.itemModal.url,
};

if (itemKey !== undefined) {
    section.items[itemKey] = item;
} else {
    section.items.push(item);
}
```

File: src/App.vue [hunk 7]
```diff
@@ -402,21 +478,60 @@ export default {
...
+        importClicked () {
+            this.isImport = true;
+            this.openImportExport();
+        },
+
+        openImportExport () {
+            this.$modal.show('transferInfos');
+
+            this.tranferTitle = this.isImport ? 'Import' : 'Export';
+            let local = window.localStorage;
```
Comment (wesleyhf) [inline src/App.vue]: smoke test on `localStorage`

File: src/App.vue [hunk 8]
```diff
@@ -402,21 +478,60 @@ export default {
...
             this.sectionModal.title = '';
         },
 
+        importClicked () {
+            this.isImport = true;
+            this.openImportExport();
+        },
+
+        openImportExport () {
+            this.$modal.show('transferInfos');
```
Comment (wesleyhf) [inline src/App.vue]: call after

File: src/App.vue (hunk 4 above)
Comment (wesleyhf) [inline src/App.vue]: check why you are using `slashes`

File: src/App.vue [hunk 9]
```diff
@@ -402,21 +478,60 @@ export default {
...
             this.sectionModal.sectionKey = null;
             this.sectionModal.title = '';
         },
 
+        importClicked () {
+            this.isImport = true;
+            this.openImportExport();
+        },
+
+        openImportExport () {
```
Comment (wesleyhf) [inline src/App.vue]: separate this function into `import and export` and populate the values on modal separated

File: src/App.vue [hunk 10]
```diff
@@ -402,21 +478,60 @@ export default {
...
+        },
+
+        submitImportExport (newStorage) {
+            this.validateText(JSON.parse(newStorage));
+
+            this.closeImportExport();
+        },
+
+        closeImportExport () {
+            this.$modal.hide('transferInfos');
```
Comment (wesleyhf) [inline src/App.vue]: call after

File: src/App.vue [hunk 11]
```diff
@@ -429,6 +544,50 @@ export default {
                 'saveProperties': ['theme', 'sections'],
             };
         },
+
+        validateText(importText) {
```
Comment (wesleyhf) [inline src/App.vue]: remove this function and put his logic on `submitImportExport`

File: src/App.vue [hunk 12]
```diff
@@ -402,21 +478,60 @@ export default {
...
+            let local = window.localStorage;
+
+            let storage = JSON.stringify(local.App).replace(/\\/g, "");
+
+            if(! this.isImport) {
+                this.transferModal.data = storage;
+            }
+        },
+
+        submitImportExport (newStorage) {
```
Comment (wesleyhf) [inline src/App.vue]: change to `importJson` or something like that

File: src/App.vue [hunk 13]
```diff
@@ -429,6 +544,50 @@ export default {
                 'saveProperties': ['theme', 'sections'],
             };
         },
+
+        validateText(importText) {
+            if (! importText.theme) {
```
Comment (wesleyhf) [inline src/App.vue]: don't return directly, because you don't check if the json has `sections` key

Comment (wesleyhf) [reply src/App.vue]: and use `indexOf` directly

File: src/App.vue [hunk 14]
```diff
@@ -429,6 +544,50 @@ export default {
...
+                return;
+            }
+
+            if(! importText.sections) {
+                return;
+            }
+
+            let sectionsToImport = [];
+
+            importText.sections.forEach(section => {
```
Comment (wesleyhf) [inline src/App.vue]: use `reduce`, if you need a help call me

File: src/App.vue [hunk 15]
```diff
@@ -429,6 +544,50 @@ export default {
...
+                if(! section.title) {
+                    return;
+                }
+
+                let itemsToImport = [];
+                section.items.forEach((item) => {
```
Comment (wesleyhf) [inline src/App.vue]: `reduce`

Comment (wesleyhf) [inline src/App.vue]: when you have only one parameter remove the parenthesis

File: src/App.vue [hunk 16]
```diff
@@ -122,130 +148,155 @@
...
+                        <button
+                            class="
+                                bg-primary hover:bg-primary-light
+                                text-white font-bold py-2 px-4 rounded
+                                border-b-4 border-primary-dark hover:border-primary
+                            "
+                            @click="submitItemModal()"
+                            v-text="'Add Item'"
+                        >
+                        </button>
```
Comment (gabrielboliveira) [inline src/App.vue]: move the `</button>` to the same line as the `>`

File: src/App.vue [hunk 17]
```diff
@@ -122,130 +148,155 @@
...
+                        <button
+                            class="
+                                bg-primary hover:bg-primary-light
+                                text-white font-bold py-2 px-4 rounded
+                                border-b-4 border-primary-dark hover:border-primary
+                            "
+                            @click="submitSectionModal()"
+                            v-text="'Add Section'"
+                        >
+                        </button>
```
Comment (gabrielboliveira) [inline src/App.vue]: Same as above

File: src/App.vue [hunk 18]
```diff
@@ -122,130 +148,155 @@
...
+                    </div>
+
+                    <div class="border-b border-b-2 border-primary py-2 mt-3">
+                        <textarea
+                            class="appearance-none bg-transparent border-none w-full text-grey-darker py-1 px-2"
+                            rows="9"
+                            v-model="importExportModal.data"
+                        >
+                        </textarea>
```
Comment (gabrielboliveira) [inline src/App.vue]: Same as above

File: src/App.vue [hunk 19]
```diff
@@ -371,50 +428,139 @@ export default {
...
+                section.items[itemKey] = item;
+            } else {
+                section.items.push(item);
+            }
 
             this.closeItemModal();
         },
 
+        removeItem(sectionKey, itemKey) {
+            if(confirm("Delete item?")) {
```
Comment (gabrielboliveira) [inline src/App.vue]: Use single quotes on js

File: src/App.vue [hunk 20]
```diff
@@ -371,50 +428,139 @@ export default {
...
+                    title: this.sectionModal.title,
+                    items: [],
+                });
+            }
 
             this.closeSectionModal();
         },
 
+        removeSection(sectionKey) {
+            if(confirm("Delete section?")) {
```
Comment (gabrielboliveira) [inline src/App.vue]: Same - single quotes

File: src/App.vue [hunk 21]
```diff
@@ -371,50 +428,139 @@ export default {
...
+                    return {
+                        title: item.title,
+                        icon: item.icon || '',
+                        image: item.image || '',
+                        url: item.url,
+                    };
+                });
+
+                return section;
+            });
```
Comment (gabrielboliveira) [inline src/App.vue]: Break the logic into new lines:

```js
let sections = importText
    .sections
    .filter(section => !! section.title)
    .map(section => {
        section.items = section
            .items
            .filter(item => item.title && item.url)
            .map(item => ({
                title: item.title,
                icon: item.icon || '',
                image: item.image || '',
                url: item.url,
            }));

        return section;
    });
```

Also take a look on some direct returns I changed, eg on the filter method

File: src/App.vue [hunk 22]
```diff
@@ -371,50 +425,138 @@ export default {
...
+                        .items
+                        .filter(item => item.title && item.url)
+                        .map(item => ({
+                            title: item.title,
+                            icon: item.icon || '',
+                            image: item.image || '',
+                            url: item.url,
+                        }));
+                return section;
+            });
```
Comment (gabrielboliveira) [inline src/App.vue]: Almost perfect. The tabulation is not right on this closing brackets and on the return. Add one tab to both

**Pull request merged**